src/compute/src/render/mod.rs

```diff
@@ -966,8 +966,90 @@ where
+                let stream = self.log_operator_hydration_inner(&oks.inner, lir_id);
+                *oks = stream.as_collection();
+            }
+            None => {
+                use ArrangementFlavor::*;
+                use SpecializedArrangement as A;
+                use SpecializedArrangementImport as AI;
+
+                let arrangement = bundle.arranged.values_mut().next().unwrap();
+                match arrangement {
+                    Local(A::RowRow(a), _) => {
+                        a.stream = self.log_operator_hydration_inner(&a.stream, lir_id);
+                    }
+                    Local(A::RowUnit(a), _) => {
+                        a.stream = self.log_operator_hydration_inner(&a.stream, lir_id);
+                    }
+                    Trace(_, AI::RowRow(a), _) => {
+                        a.stream = self.log_operator_hydration_inner(&a.stream, lir_id);
+                    }
+                    Trace(_, AI::RowUnit(a), _) => {
+                        a.stream = self.log_operator_hydration_inner(&a.stream, lir_id);
+                    }
+                }
+            }
         }
     }
+
+    fn log_operator_hydration_inner<D>(&self, stream: &Stream<G, D>, lir_id: u64) -> Stream<G, D>
+    where
+        D: Clone + 'static,
+    {
+        // Convert the dataflow as-of into a frontier we can compare with input frontiers.
+        //
+        // We (somewhat arbitrarily) define operators in iterative scopes to be hydrated when their
+        // frontier advances to an outer time that's greater than the `as_of`. Comparing
+        // `refine(as_of) < input_frontier` would find the moment when the first iteration was
+        // complete, which is not what we want. We want `refine(as_of + 1) <= input_frontier`
+        // instead.
+        let mut hydration_frontier = Antichain::new();
+        for time in self.as_of_frontier.iter() {
+            if let Some(time) = time.try_step_forward() {
+                hydration_frontier.insert(Refines::to_inner(time));
+            }
+        }
+
+        let logger = self.hydration_logger.clone();
+
+        let name = format!("LogOperationHydration ({lir_id})");
+        stream.unary_frontier(Pipeline, &name, |_cap, _info| {
+            let mut hydrated = false;
+            logger.log(lir_id, hydrated);
+
+            let mut buffer = Vec::new();
+            move |input, output| {
+                // Pass through inputs.
+                input.for_each(|cap, data| {
+                    data.swap(&mut buffer);
+                    output.session(&cap).give_vec(&mut buffer);
+                });
+
+                if hydrated {
+                    return;
+                }
+
+                let frontier = input.frontier().frontier();
+                if PartialOrder::less_equal(&hydration_frontier.borrow(), &frontier) {
+                    hydrated = true;
+                    logger.log(lir_id, hydrated);
+                }
```

Comment:
I think a different approach could be to use `RenderTimestamp::step_back` here. For `mz_repr::Timestamp`, this should be roughly the same complexity, but doesn't suffer from the ambiguities of `try_step_forward`, for WMR it might be a bit more effort, but that seems acceptable. And, I think it would make the logic clearer and wouldn't need the comment above.

Reply:
Hm, I think that wouldn't work with `as_of = [0]`. If you `step_back` time `0`, you end up with `0` again, so we would think the operator is immediately hydrated even though it's not.

I had this code at one point:

```rust
                let inner_frontier = input.frontier().frontier();
                let outer_frontier =
                    Antichain::from_iter(inner_frontier.iter().map(|t| t.clone().to_outer()));
                if PartialOrder::less_than(&as_of, &outer_frontier) {
                    hydrated = true;
                    logger.log(lir_id, hydrated);
                }
```

... but then decided to delete it because allocating a new `Antichain` on each operator invocation seems expensive.

We could get around this if we weren't comparing `Antichain`s but just timestamps, which we could do if we assumed a total order. But I don't think we can assume that with `Product` times.

Reply:
Thank you!